### PR TITLE
feat: Alertmanager webhook integration

### LINKED ISSUE #2: feat: Alertmanager webhook integration
## Status

**Implementation submitted:** https://github.com/rophy/aqsh/pull/3

---

## Use Case

When Prometheus detects an issue (high memory, high latency, disk full, etc.), the typical response involves someone noticing the alert and manually running a remediation script. This is slow and error-prone.

An Alertmanager webhook integration allows aqsh to **automatically execute task scripts in response to Prometheus alerts**, enabling automated incident response workflows.

### Webhook Endpoint

`POST /webhooks/alertmanager` — receives the standard [Alertmanager webhook payload](https://prometheus.io/docs/alerting/latest/configuration/#webhook_config) and enqueues tasks for each alert.

### Task Name Resolution

The endpoint resolves which task to run using this lookup order:

1. `alert.labels.aqsh_task` — explicit per-alert label (recommended)
2. `commonLabels.aqsh_task` — shared across the alert group
3. `alert.labels.alertname` — fallback: uses the alert name as the task name

This means you can either set an explicit `aqsh_task` label on your alert rules for fine-grained routing, or just name your tasks after your alerts.

### Alert Context via Environment Variables

Alert details are passed to scripts as env vars so scripts can make decisions based on alert context:

| Variable | Example |
|----------|---------|
| `ALERT_STATUS` | `firing` / `resolved` |
| `ALERT_NAME` | `HighMemory` |
| `ALERT_SEVERITY` | `critical` |
| `ALERT_INSTANCE` | `app-server:8080` |
| `ALERT_FINGERPRINT` | `abc123...` |
| `ALERT_LABELS_JSON` | Full labels as JSON |
| `ALERT_ANNOTATIONS_JSON` | Full annotations as JSON |
| `ALERT_LABEL_<KEY>` | Individual labels (uppercased) |
| `ALERT_ANNOTATION_<KEY>` | Individual annotations (uppercased) |

### Example Alert Rule

```yaml
groups:
  - name: aqsh-alerts
    rules:
      - alert: HighMemory
        expr: node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes < 0.1
        for: 5m
        labels:
          severity: critical
          aqsh_task: alert-handler    # ← routes to "alert-handler" task
        annotations:
          summary: "Memory usage above 90%"
```

### Example Alertmanager Config

```yaml
route:
  receiver: aqsh-webhook
  group_by: ['alertname']

receivers:
  - name: aqsh-webhook
    webhook_configs:
      - url: 'http://aqsh:8080/webhooks/alertmanager'
        send_resolved: true
```

## Open Questions

These are areas where a different approach might be preferred:

- **Endpoint path**: Is `/webhooks/alertmanager` the right convention, or would something like `/hooks/alertmanager` or `/alerts` be preferred?
- **Task name resolution**: The current `aqsh_task` label → `alertname` fallback works, but should there be a config-based mapping instead?
- **Multi-alert handling**: Currently each alert in a webhook group is enqueued as a separate task. Should there be an option to batch them?
- **Resolved alerts**: Should `resolved` alerts be forwarded to scripts by default, or should there be a config option to filter them?

### PR DESCRIPTION
**Pull request by yangminglintw**
## Summary

Adds `POST /webhooks/alertmanager` endpoint that receives [Alertmanager webhook payloads](https://prometheus.io/docs/alerting/latest/configuration/#webhook_config) and automatically enqueues aqsh tasks for each alert.

- **Task routing**: resolves task name via `aqsh_task` label → `commonLabels.aqsh_task` → `alertname` fallback
- **Alert context**: maps alert labels/annotations to `ALERT_*` environment variables for scripts
- **Multi-alert handling**: each alert in a group is enqueued as a separate task

## Changes (7 files, +436 lines)

| File | Description |
|------|-------------|
| `internal/webhook/alertmanager.go` | Payload structs, `ResolveTaskName()`, `AlertToEnv()` |
| `internal/webhook/alertmanager_test.go` | 11 unit tests covering routing, env mapping, edge cases |
| `internal/api/api.go` | `handleAlertmanagerWebhook()` handler + route registration |
| `docs/api.md` | Webhook endpoint documentation |
| `tasks.yaml` | `alert-handler` + `latency-handler` task definitions |
| `tasks/alert-handler.sh` | Demo remediation script |
| `tasks/latency-handler.sh` | Demo latency handler script |

## Testing

- 11 new unit tests for the webhook package
- All 44 existing + new tests pass (`go test ./...`)

## Demo

The included `alert-handler.sh` and `latency-handler.sh` scripts demonstrate how alert context flows into task scripts via environment variables:

```bash
# alert-handler.sh receives:
# ALERT_STATUS=firing, ALERT_NAME=HighMemory, ALERT_SEVERITY=critical, etc.
```

## Example Alertmanager config

```yaml
receivers:
  - name: aqsh-webhook
    webhook_configs:
      - url: 'http://aqsh:8080/webhooks/alertmanager'
        send_resolved: true
```

Closes #2

## Summary by CodeRabbit

* **New Features**
  * Alertmanager webhook integration to create tasks from incoming alerts.
  * New alert and latency handler tasks and accompanying remediation scripts.

* **Documentation**
  * Added API docs for the Alertmanager webhook with request/response examples and env mappings.

* **Tests**
  * Unit tests covering alert-to-task name resolution and environment variable generation.

### CODE REVIEW

File: internal/api/api.go [hunk 1]
```diff
@@ -282,6 +284,93 @@ func (s *Server) handleGetLogs(w http.ResponseWriter, r *http.Request) {
...
+	var wh webhook.AlertmanagerWebhook
+	if err := json.NewDecoder(r.Body).Decode(&wh); err != nil {
+		s.jsonError(w, http.StatusBadRequest, "invalid JSON: "+err.Error())
+		return
+	}
+
+	if len(wh.Alerts) == 0 {
+		s.jsonError(w, http.StatusBadRequest, "no alerts in payload")
+		return
+	}
```
[the remark this answers is not in the document]
Comment (yangminglintw) [reply internal/api/api.go:297]: Good point. aqsh is designed to sit behind a reverse proxy (e.g., nginx, Traefik) that handles authentication — same as handleSubmitTask relies on identity/group headers set by the proxy. Adding a webhook secret would be a nice defense-in-depth option though. Tracking as a follow-up.

File: internal/api/api.go [hunk 2]
```diff
@@ -282,6 +284,93 @@ func (s *Server) handleGetLogs(w http.ResponseWriter, r *http.Request) {
...
+			TaskID:           info.ID,
+			TaskName:         taskName,
+			Status:           "pending",
+		})
+	}
+
+	if successCount == 0 {
+		s.jsonResponse(w, http.StatusBadRequest, map[string]any{"results": results})
+		return
+	}
```
[the remark this answers is not in the document]
Comment (yangminglintw) [reply internal/api/api.go:377]: Acknowledged. For the initial implementation, the current behavior (400 when all alerts fail, 202 when any succeed) keeps things simple. In practice, config errors (unknown task) are far more common than Redis outages. Will consider adding 503 distinction if Redis transient failures become an issue in production.
